[DOC] Add ingest processors documentation (#4299)
opensearch-project/documentation-website: _api-reference/ingest-apis/simulate-ingest.md

---
layout: default
title: Simulate pipeline
parent: Ingest pipelines
grand_parent: Ingest APIs
nav_order: 11
redirect_from:
  - /opensearch/rest-api/ingest-apis/simulate-ingest/
---

# Simulate pipeline

Use the simulate ingest pipeline API operation to run or test the pipeline.

## Path and HTTP methods

The following requests **simulate the latest ingest pipeline created**:

```
GET _ingest/pipeline/_simulate
POST _ingest/pipeline/_simulate
```
{% include copy-curl.html %}

The following requests **simulate a single pipeline based on the pipeline ID**:

```
GET _ingest/pipeline/<pipeline-id>/_simulate
POST _ingest/pipeline/<pipeline-id>/_simulate
```
{% include copy-curl.html %}

## Request body fields

The following table lists the request body fields used to run a pipeline.

Field | Required | Type | Description
:--- | :--- | :--- | :---
`docs` | Required | Array | The documents to be used to test the pipeline.
`pipeline` | Optional | Object | The pipeline to be simulated. If the pipeline identifier is not included, then the response simulates the latest pipeline created.

> The `docs` field can include subfields listed in the following table.

Reviewer: ```suggestion
The `docs` field can include the subfields listed in the following table.
```

Field | Required | Type | Description
:--- | :--- | :--- | :---
`source` | Required | Object | The document's JSON body.
`id` | Optional | String | A unique document identifier. The identifier cannot be used elsewhere in the index.

> `index` | Optional | String | The index where the document's transformed data appears.

Reviewer: ```suggestion
`index` | Optional | String | The index in which the document's transformed data appears.
```

#### Example: Specify a pipeline in the path

```json
POST /_ingest/pipeline/my-pipeline/_simulate
{
  "docs": [
    {
      "_index": "my-index",
      "_id": "1",
      "_source": {
        "grad_year": 2024,
        "graduated": false,
        "name": "John Doe"
      }
    },
    {
      "_index": "my-index",
      "_id": "2",
      "_source": {
        "grad_year": 2025,
        "graduated": false,
        "name": "Jane Doe"
      }
    }
  ]
}
```
{% include copy-curl.html %}

The request returns the following response:

```json
{
  "docs": [
    {
      "doc": {
        "_index": "my-index",
        "_id": "1",
        "_source": {
          "name": "JOHN DOE",
          "grad_year": 2023,
          "graduated": true
        },
        "_ingest": {
          "timestamp": "2023-06-20T23:19:54.635306588Z"
        }
      }
    },
    {
      "doc": {
        "_index": "my-index",
        "_id": "2",
        "_source": {
          "name": "JANE DOE",
          "grad_year": 2023,
          "graduated": true
        },
        "_ingest": {
          "timestamp": "2023-06-20T23:19:54.635746046Z"
        }
      }
    }
  ]
}
```

### Example: Verbose mode

> When the previous request is run with the `verbose` parameter set to `true`, the response shows the sequence of transformations made on each document. For example, for the document with the ID `1`, the response contains the results of applying each processor in the pipeline in turn:

Reviewer: ```suggestion
When the previous request is run with the `verbose` parameter set to `true`, the response shows the sequence of transformations for each document. For example, for the document with the ID `1`, the response contains the results of applying each processor in the pipeline in sequence:
```

              "pipeline": "my-pipeline",
              "timestamp": "2023-06-20T23:23:26.656564631Z"
            }
          }
        },
        {
          "processor_type": "set",
          "status": "success",
          "description": "Sets 'graduated' to true",
          "doc": {
            "_index": "my-index",
            "_id": "1",
            "_source": {
              "name": "John Doe",
              "grad_year": 2023,
              "graduated": true
            },
            "_ingest": {
              "pipeline": "my-pipeline",
              "timestamp": "2023-06-20T23:23:26.656564631Z"
            }
          }
        },
        {
          "processor_type": "uppercase",
          "status": "success",
          "doc": {
            "_index": "my-index",
            "_id": "1",
            "_source": {
              "name": "JOHN DOE",
              "grad_year": 2023,
              "graduated": true
            },
            "_ingest": {
              "pipeline": "my-pipeline",
              "timestamp": "2023-06-20T23:23:26.656564631Z"
            }
          }
        }
      ]
    }
  ]
}
```

### Example: Specify a pipeline in the request body

> Alternatively, you can specify a pipeline directly in the request body without creating a pipeline first:

Reviewer: ```suggestion
Alternatively, you can specify a pipeline directly in the request body without first creating a pipeline:
```
